File: reference/dotnet/Adapters/Company.Product.Adapters.Rest.Generated/Models/AnimalMapper.cs
namespace Company.Product.Adapters.Rest.Models;

public class AnimalMapper : Domain.UseCases.Types.AnimalVisitor<Animal>
{
    public static Animal FromDomain(Domain.UseCases.Types.Animal animal)
    {
        return animal.Visit(new AnimalMapper());
    }

    public static Domain.UseCases.Types.Animal ToDomain(Animal animal)
    {
        return animal.ToDomain();
    }

    public Animal VisitCat(Domain.UseCases.Types.Cat cat)
    {
        return new Cat()
        {
            AnimalId = cat.AnimalId,
            Sound = cat.Sound,
            B = cat.B
        };
    }

    public Animal VisitCow(Domain.UseCases.Types.Cow cow)
    {
        return new Cow()
        {
            AnimalId = cow.AnimalId,
            Sound = cow.Sound,
            C = cow.C
        };
    }

    public Animal VisitDog(Domain.UseCases.Types.Dog dog)
    {
        return new Dog()
        {
            AnimalId = dog.AnimalId,
            Sound = dog.Sound,
            A = dog.A
        };
    }
}

File: reference/springboot/adapters-rest-generated/src/main/java/org/example/adapters/rest/AnimalMapper.java
package org.example.adapters.rest;

import org.example.adapters.rest.dto.AnimalDto;
import org.example.adapters.rest.dto.CatDto;
import org.example.adapters.rest.dto.CowDto;
import org.example.adapters.rest.dto.DogDto;
import org.example.domain.usecases.types.*;

public class AnimalMapper {

    public static AnimalDto fromDomain(Animal domain) {
        return domain.visit(new AnimalVisitor<>() {

            @Override
            public AnimalDto visitDog(Dog dog) {
                var dto = new DogDto();
                dto.setA(dog.getA());
                setAnimal(dto, dog);
                return dto;
            }

            @Override
            public AnimalDto visitCat(Cat cat) {
                var dto = new CatDto();
                dto.setB(cat.getB());
                setAnimal(dto, cat);
                return dto;
            }

            @Override
            public AnimalDto visitCow(Cow cow) {
                var dto = new CowDto();
                dto.setC(cow.getC());
                setAnimal(dto, cow);
                return dto;
            }

            private void setAnimal(AnimalDto dto, Animal animal) {
                dto.setAnimalId(animal.getAnimalId());
                dto.setSound(animal.getSound());
            }
        });
    }

    public static Animal toDomain(AnimalDto dto) {
        return dto.toDomain();
    }
}
